eclass/kernel-build.eclass

```diff
@@ -534,40 +538,85 @@ kernel-build_src_install() {
 			dracut "${dracut_args[@]}" "${image%/*}/initrd" ||
 				die "Failed to generate initramfs"
 
-			local ukify_args=(
+			local ukify_base_args=(
 				--linux="${image}"
 				--initrd="${image%/*}/initrd"
-				--cmdline="${KERNEL_GENERIC_UKI_CMDLINE}"
 				--uname="${KV_FULL}"
-				--output="${image%/*}/uki.efi"
 			)
 
-			if [[ ${KERNEL_IUSE_MODULES_SIGN} ]] && use secureboot; then
-				ukify_args+=(
-					--signtool=sbsign
-					--secureboot-private-key="${SECUREBOOT_SIGN_KEY}"
-					--secureboot-certificate="${SECUREBOOT_SIGN_CERT}"
+			local pcrpkey=${ED}${kernel_dir}/certs/signing_key.x509
+			if [[ -r ${pcrpkey} ]]; then
+				ukify_base_args+=(
+					--pcrpkey="${pcrpkey}"
 				)
-				if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
-					ukify_args+=(
-						--signing-engine="pkcs11"
-					)
-				else
-					# Sytemd-measure does not currently support pkcs11
-					ukify_args+=(
+			fi
+
+			ukify build "${ukify_base_args[@]}" \
+				--output="${T}/base-uki.efi" ||
+					die "Failed to generate base UKI"
+
+
+			local ukify_extend_args=()
+			if [[ ${KERNEL_IUSE_MODULES_SIGN} ]] && use secureboot; then
+				ukify_extend_args+=(
 						--measure
-						--pcrpkey="${ED}${kernel_dir}/certs/signing_key.x509"
 						--pcr-private-key="${SECUREBOOT_SIGN_KEY}"
 						--phases="enter-initrd"
 						--pcr-private-key="${SECUREBOOT_SIGN_KEY}"
 						--phases="enter-initrd:leave-initrd enter-initrd:leave-initrd:sysinit enter-initrd:leave-initrd:sysinit:ready"
+				)
+				if [[ ${SECUREBOOT_SIGN_KEY} == pkcs11:* ]]; then
+					ukify_extend_args+=(
+						--signing-engine="pkcs11"
+						--pcr-public-key="${SECUREBOOT_SIGN_CERT}"
 					)
 				fi
 			fi
 
-			# systemd<255 does not install ukify in /usr/bin
-			PATH="${PATH}:${BROOT}/usr/lib/systemd:${BROOT}/lib/systemd" \
-				ukify build "${ukify_args[@]}" || die "Failed to generate UKI"
+			# Note, we cannot use an associative array here because those are
+			# not ordered.
+			local profiles=()
+			local cmdlines=()
+
+			# If defined, make the custom entry the first and default
+			if [[ -n ${KERNEL_GENERIC_UKI_CMDLINE} ]]; then
+				profiles+=(
```

Comment:
> Can we use an associative array for that?

Unfortunately not, associative arrays are inherently unordered. And we do want to define the ordering (i.e. custom first if it's set, otherwise the current in tree default is first so there is no change sudden in behaviour).

That being said, I do agree this looks a bit weird, I have not yet found a way to make it more clean.